R/collapseHz.R

```diff
@@ -0,0 +1,279 @@
+                                      } else {
+                                        v
+                                      }
+                                    },
+                                    top = .SD[[hzd[1]]],
+                                    bottom = .SD[[hzd[2]]]),
+                         
+                           # process custom aggregation functions (may return data.frames)
+                           do.call('c', lapply(colnames(.SD)[colnames(.SD) %in% naf], 
+                                               function(n, top, bottom) {
+                                                         out <- AGGFUN[[n]](.SD[[n]], top, bottom)
+                                                         if (!is.data.frame(out)) {
+                                                           out <- data.frame(out)
+                                                           colnames(out) <- n
+                                                         } else {
+                                                           colnames(out) <- paste0(n, ".", colnames(out))
+                                                         }
+                                                         out
+                                                                           },
+                                               top = .SD[[hzd[1]]], 
+                                               bottom = .SD[[hzd[2]]]))), 
+                 by = g[gidx]]
+        # remove grouping ID
+        res$g <- NULL
+      } else {
+        res <- h[0, ]
+      }
+      
+      # allow for replacing values as well as adding new values with data.frame AGGFUN
+      test1.idx <- na.omit(match(colnames(res), paste0(naf, ".", naf))) 
+      test2.idx <- na.omit(match(paste0(naf, ".", naf), colnames(res)))
+      colnames(res)[test2.idx] <- naf[test1.idx]
+      
+      # determine matches that are only a single layer (no aggregation applied)      
+      res2 <- h[hidx & l, ]
+      res2$hzdeptnew <- res2[[hzd[1]]]
+      res2$hzdepbnew <- res2[[hzd[2]]]
+      res2[[hzd[1]]] <- NULL
+      res2[[hzd[2]]] <- NULL
+      
+      # combine matches
+      res3 <- data.table::rbindlist(list(res, res2), fill = TRUE)
+      if (missing(by) && nrow(res3) > 0){
+        res3[[hzdesgn]] <- labels[p]
+      }
+      
+      # combine matches with horizons that did not match
+      agg.idx <- which(g %in% unique(g[l]) | hidx)
+      if (length(agg.idx) > 0) {
+        h <- h[-agg.idx, ]
+      }
+      h <- data.table::rbindlist(list(h, res3), fill = TRUE)
+      
+      # replace depths
+      hn <- !is.na(h$hzdeptnew) & !is.na(h$hzdepbnew)
+      h[[hzd[1]]][hn] <- h$hzdeptnew[hn]
+      h[[hzd[2]]][hn] <- h$hzdepbnew[hn]
+      h$hzdeptnew <- NULL
+      h$hzdepbnew <- NULL
+      
+      # sort horizons by id name and top depth
+      h <- h[order(h[[idn]], h[[hzd[1]]]),]
+      
+      # replace horizons in parent SPC
+      replaceHorizons(x) <- h
```

Comment:
Should we retain original horizon IDs and set new ones? This level of accounting could be useful, but adds clutter. I'd vote for new horizon IDs and dump the old ones for simplicity.

Reply:
> Should we retain original horizon IDs and set new ones? This level of accounting could be useful, but adds clutter. 
I don't think it is a good idea to dump the horizon IDs and calculate new ones. This could create unintended conflicts with the auto-calculated sequential integer IDs.

I think it not desirable for the calculated `hzID` nor one set by the user e.g. `chkey` or `chiid`. Essentially I want to retain reference to the source horizon. I don't see an issue with current behavior that keeps a subset of the parent `hzidcolumn()` values. Can you elaborate on what you are thinking here?

Reply:
Maybe I missed something: when two horizons are *collapsed* what is the resulting horizon ID?

Reply:
The horizon ID (character data type) should be from the thickest horizon in the group. It is not handled differently than any other horizon-level variable by default, but you can supply a custom aggregation function (to e.g. concatenate IDs) if desired